### DISCUSSION TITLE
How to add multiple rules with click of Add Rule button?

### DISCUSSION THREAD

**Discussion by kellywang1234** [react-querybuilder/react-querybuilder#604] [Q&A]
Hi,
Do you know how to add two or multiple rows at a time when we click the Add Rule button?
It would be really helpful if there is an example! Thank you very much!
This is sandbox link if you click AddRule (Number), Weight and Height rows should display at the same time https://codesandbox.io/s/react-query-builder-additional-rule-button-for-different-fields-and-operators-nspxw3?file=/src/App.tsx:1019-1026

**Comment by jakeboone02** [comment]
This is definitely possible, but it's a little tricky up through the current release, v6. If you can wait for v7, this will be a _much_ easier question to answer, but I'm not exactly sure when v7 final will be released.

_(Feel free to use the [latest alpha, 7.0.0-alpha.2](https://www.npmjs.com/package/react-querybuilder/v/7.0.0-alpha.2), if you feel adventurous ([migration instructions here](https://react-querybuilder.js.org/docs/next/migrate)). Honestly it's probably safe to use in production, and I'm not planning on making any big changes before the final release. I'm happy to help you migrate to v7 and use the new features that make your requirement easier to implement, but if you need something "official", read on.)_

If you only need to deal with one level of rules, i.e. no sub-groups, then it's probably easiest to use the techniques described in the [External Controls](https://react-querybuilder.js.org/docs/tips/external-controls) page.

If you need to handle sub-groups properly, you'll need to move the state management of the query outside the component. In the `addRuleAction` component, call whatever update method is appropriate (e.g. `dispatch(actionCreator(...))` for Redux) instead of `props.handleOnClick`. Pass the `path` and whatever `context` you need to utilize the [`add` method](https://react-querybuilder.js.org/docs/utils/misc#add).

In #296 I linked to [this CodeSandbox](https://codesandbox.io/s/59x264) which does something similar. Hopefully that gives you a starting point at least.

**Reply by kuljeet2604** [reply]
can we modify it to add independent rule group instead of nested?

**Reply by jakeboone02** [reply]
You might want to try the `showCloneButtons` prop ([documentation](https://react-querybuilder.js.org/docs/components/querybuilder#showclonebuttons)). That will display a button to duplicate the current group (or rule) as a sibling object, not nested.
